test/OperationTest.cpp

```diff
@@ -134,3 +138,128 @@ TEST_F(OperationTestFixture, verifyCachePreventsInProgressState) {
+class ShallowParentOperation : public Operation {
+  std::shared_ptr<QueryExecutionTree> child_;
+
+  explicit ShallowParentOperation(std::shared_ptr<QueryExecutionTree> child)
+      : child_{std::move(child)} {}
+  string asStringImpl([[maybe_unused]] size_t) const override {
+    return "ParentOperation";
+  }
+  string getDescriptor() const override { return "ParentOperationDescriptor"; }
+  size_t getResultWidth() const override { return 0; }
+  size_t getCostEstimate() override { return 0; }
+  uint64_t getSizeEstimateBeforeLimit() override { return 0; }
+  float getMultiplicity([[maybe_unused]] size_t) override { return 0; }
+  bool knownEmptyResult() override { return false; }
+  vector<ColumnIndex> resultSortedOn() const override { return {}; }
+  VariableToColumnMap computeVariableToColumnMap() const override { return {}; }
+
+ public:
+  template <typename ChildOperation, typename... Args>
+  static ShallowParentOperation of(QueryExecutionContext* qec, Args&&... args) {
+    return ShallowParentOperation{
+        ad_utility::makeExecutionTree<ChildOperation>(qec, args...)};
+  }
+
+  std::vector<QueryExecutionTree*> getChildren() override {
+    return {child_.get()};
+  }
+
+  ResultTable computeResult() override {
+    auto childResult = child_->getResult();
+    return {childResult->idTable().clone(), resultSortedOn(),
+            childResult->getSharedLocalVocab()};
+  }
+
+  // Provide public view of remainingTime for tests
+  std::chrono::milliseconds publicRemainingTime() const {
+    return remainingTime();
+  }
+};
+
+template <>
+void QueryExecutionTree::setOperation<StallForeverOperation>(
+    std::shared_ptr<StallForeverOperation> operation) {
+  _rootOperation = std::move(operation);
+}
+
+// _____________________________________________________________________________
+
+TEST(OperationTest, verifyExceptionIsThrownOnCancellation) {
+  auto qec = getQec();
+  auto handle = std::make_shared<CancellationHandle>();
+  ShallowParentOperation operation =
+      ShallowParentOperation::of<StallForeverOperation>(qec);
+  operation.recursivelySetCancellationHandle(handle);
+
+  std::thread thread{[&]() {
+    std::this_thread::sleep_for(5ms);
+    handle->cancel(CancellationState::TIMEOUT);
+  }};
+  EXPECT_THROW(
+      {
+        try {
+          operation.computeResult();
+        } catch (const ad_utility::AbortException& exception) {
+          EXPECT_THAT(exception.what(),
+                      ::testing::HasSubstr("Cancelled due to timeout"));
+          throw;
+        }
+      },
+      ad_utility::AbortException);
+  thread.join();
+}
+
+// _____________________________________________________________________________
+
+TEST(OperationTest, verifyRemainingTimeDoesCountDown) {
+  auto qec = getQec();
+  ShallowParentOperation operation =
+      ShallowParentOperation::of<StallForeverOperation>(qec);
+  operation.recursivelySetTimeConstraint(1ms);
+  std::this_thread::sleep_for(1ms);
```

Comment:
You should also check that the remaining time is larger than 0ms before the sleep.